### ISSUE TITLE
Move off Foundation

### ISSUE THREAD

**Issue by drewcrawford** [AnarchyTools/atbuild#81]
@AnarchyTools/build-tool-admins @AnarchyTools/package-manager-admins 

So, having done another round of "Foundation change of the week", I have lost all confidence in Foundation as a library.  There are too many differences between Swift 2, Swift 3, Corelibs Foundation and Darwin Foundation, we have hundreds of lines of code working around "platform X" issues and I hereby declare Foundation bankruptcy.

I intend to eventually migrate atpkg / atbuild off Foundation, and the only question in my mind is what to migrate them to.  But it should be a library that's the same everywhere and very boring.  You know, like our philosophy.

I think there are 3 options.
1.  [SwiftFoundation](https://github.com/PureSwift/SwiftFoundation) is the obvious candidate.  However, they're not packaged for us, and while their code is okay, I compared it to some other "reimplement a Foundation class" code I have lying around, and I believe we can do better.
   
   We could do all that and send it back to them, although I don't know how receptive they will be to our changes.  If this makes a difference we could ping them to this issue and see if they respond.
2.  Alternatively, we could write our own Foundation library, perhaps lifting SwiftFoundation code where we don't want to do better, and our projects take a dependency on that.  This is my personal preference, but I don't know what the temperature is on starting "yet another anarchy project".
3.  I could go off and do my own thing, and we link it from atpkg and atbuild.  atpm is free to make its own decision, but I bet it would want to use my code instead of upstream's.  This option is best if you guys don't have the bandwidth to be involved.  I already have written around 10% of the code we need spread out between other projects.

To be clear, I'm not suggesting we do this overnight (although that does seem to be the way we implement most of our projects :smile:  ) but over a few months.  Specifically, several of the pieces I would like to have at the start are not in place:
- [ ]  To implement a new dependency across atbuild and atpkg, we really need a tool like `atpm` to wrangle the dependency chain for us
- [ ]  I don't run `atpm` right now, primarily because of the static linking issue, and my desire to not keep yet another binary up to date on all platforms all the time.
- [ ]  We need to get the static build stuff merged
- [ ] Alternatively we could start distributing Swift ourselves and merge in my changes out of tree.  I already _practically_ do this for Linux, it's a question of the additional effort to also build/distribute OSX.

Would appreciate any thoughts you may have.  I've just reached my snapping point with Foundation and I'm trying to plan a smooth migration to something that doesn't surprise us 3 times a month.

**Comment by owensd** [comment]
So this is something that I've thought about from the start. I don't like Swift Foundation, I don't like the idea behind it, and I don't think it's long-term the right solution. I don't think the ObjC APIs are a good fit for Swift, at all.

So what's the intention? Re-implement the Foundation APIs as they, or build an API set that's equivalent to Foundation but actually built for Swift?

If it's to re-implement Foundation, I think we ought to check out SwiftFoundation and get their temperature on it and get them building with our toolset as well. 

If it's to just write the APIs we need, then option 2 is what I'd vote for, the caveat that it's mostly for internal consumption and not really something we are trying to push out to others.

The option is to fork Foundation and make the fixes we need there with the hopes that we can merge some of it or at least they'll fix it up over time...

**Comment by drewcrawford** [comment]
So let's separate out three things:
1.  Reimplement Foundation API-for-API
2.  Design APIs for Swift
3.  2, but only the APIs we need for atbuild and atpkg

I have no interest in 1.  So that leaves 2 and 3.

The problem with 3 is that atbuild is not the first project in my list to declare Foundation bankrupcty.  So I already have accumulated, independent of whatever we decide:
- [x] Loading / saving strings to disk
- [x] Regular expressions
- [x] Some file manipulation

atbuild / atpkg would need additionally:
- [ ] More file manipulation
- [ ] String manipulation

But given that some of the code we need is already in _other invisible trees_, I think it would make sense to set up a shared repositor[ies] so that fixes from my internal stuff can flow into atbuild and vice versa.  And "string manipulation" and "file manipulation" are just problems that come up a lot, I think we would be well-served to solve them at a library level that is prepared to serve many projects.

That "library level" might very well be SwiftFoundation, but I don't know how receptive they would be to our changes.  Off the top of my head:
1.  They'd need a build.atpkg
2.  Separately from our problems, I need to build at least a few components on Swift 2 & 3, and they are 3-only
3.  They seem stable and I would like to improve their APIs in some cases.  For example, [here is their regex library](https://github.com/PureSwift/SwiftFoundation/blob/develop/Sources/SwiftFoundation/POSIXRegularExpression.swift).  I have in my pocket one that is IMO quite a lot better, but I have no idea what their view of it would be or if they would even entertain those kinds of changes.

Alternatively, that "library level" might be a project under my namespace, I am fine with that too.  I don't want to push new work onto you guys that you don't want to be responsible for.

**Comment by owensd** [comment]
Personally, I'd like to minimize our dependencies across the entire stack. So putting everything we actually need and are building under AnarchyTools is fine by me.

If SwiftFoundation doesn't look promising or has limitations we cannot accept (e.g. no Swift v2 support), I'm all for not pursuing it. 

At the end of the day, I'd really like to be able just point to the Anarchy Tools github and say, "look, this is everything we need to do in order to actually build software. Please help us reduce this and actually make it part of Swift."

**Comment by drewcrawford** [comment]
>  I'd really like to be able just point to the Anarchy Tools github and say, "look, this is everything we need to do in order to actually build software. Please help us reduce this and actually make it part of Swift."

Working back from this goal, I think the easiest way to achieve it is to define a few microlibraries with obvious upstream targets.
- "Standard Library Plus" which contains things that obviously should be in the core standard library but are not for some reason.  By this I mean things like `String.hasSuffix`, `componentsSeparatedbyString` etc.  Only obvious extensions on the standard library types, and maybe one or two new "why isn't this a thing" types like `Data` to work with bytes.
- A file I/O library with your basic copy, save, load, mkdir, etc.  Here we would target NSFileManager.

I think that is probably all atbuild/atpkg needs, there may some minor things I'm missing.

Going forward, I have internal projects that are reimplementing parts of Foundation beyond NSFileManager that I would be happy to donate somewhere, but I don't particularly care if we are the ones who take them or not.

**Comment by dunkelstern** [comment]
So to chime in and complete this thread:

I already had String, Date and File classes in my Unchained Web Framework class, so I implemented a set of sane APIs and packaged it into `atFoundation`.

We currently have the following:
- `FS` class (roughly equivalent to NSFilemanager but with a Swifty API)
- `Path` struct for working with filesystem and URL paths
- `FileInfo` struct for `stat`-type information of files
- `Charset` class for working with unicode (`isLetter`, `isNumberDigit`, etc.)
- `File` class that allows reliable reading/writing/copying, etc.
- `SysError` enum that encapsulates `errno` error codes into a unified interface as the `errno` codes differ on different platforms
- `URL` struct for parsing and generating URLs
- A lot of extensions to `String` (replace sub-string, find sub-string/character, split, join, sub-string via range/index, whitespace trimming, URL-encoding/decoding)

It seems this will suffice for porting all anarchy tools over to it. The library is heavily tested too, but coverage is not quite ideal.

I started with `atpkg` already but it bugs out currently. Grabbing this ticket for the other tools too.
